File: 02_Data_Structures/trees/trie/implementations/operations.py
# File: operations.py
# Topic: Trie (Prefix Tree) Operations in Python
#
# Implements core trie functions:
#   - insert(word)
#   - search(word)
#   - starts_with(prefix)
#   - delete(word) (optional advanced)
#
# Time Complexity per operation:
#   Insert: O(L)
#   Search: O(L)
#   Prefix: O(L)
# Space Complexity: O(N * alphabet_size)
# where L = length of the word and N = number of stored words.

class TrieNode:
    def __init__(self):
        self.children = {}
        self.end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    # Insert a word into the trie
    def insert(self, word):
        cur = self.root
        for ch in word:
            if ch not in cur.children:
                cur.children[ch] = TrieNode()
            cur = cur.children[ch]
        cur.end = True

    # Check if word exists
    def search(self, word):
        cur = self.root
        for ch in word:
            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return cur.end

    # Check if any word begins with prefix
    def starts_with(self, prefix):
        cur = self.root
        for ch in prefix:
            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return True

    # Optional: delete a word safely
    def delete(self, word):
        return self._delete(self.root, word, 0)

    # Helper to delete nodes only when redundant
    def _delete(self, node, word, idx):
        if idx == len(word):
            if not node.end:
                return False
            node.end = False
            return len(node.children) == 0

        ch = word[idx]
        if ch not in node.children:
            return False

        should_delete = self._delete(node.children[ch], word, idx + 1)

        if should_delete:
            del node.children[ch]
            return not node.end and len(node.children) == 0

        return False


# Example usage:
if __name__ == "__main__":
    t = Trie()
    t.insert("apple")
    t.insert("app")

    print(t.search("apple"))    # True
    print(t.search("app"))      # True
    print(t.search("ap"))       # False
    print(t.starts_with("ap"))  # True

File: 02_Data_Structures/trees/trie/implementations/operations.cpp
// File: operations.cpp
// Topic: Trie (Prefix Tree) Implementation
//
// This file contains core Trie operations:
// 1. insert(word)
// 2. search(word)
// 3. startsWith(prefix)
//
// A trie efficiently stores strings for prefix queries.
// Time complexities:
//   Insert:   O(L)
//   Search:   O(L)
//   Prefix:   O(L)
// where L = length of the word.
//
// Common interview applications:
//   - Autocomplete
//   - Spell-check
//   - Word existence check
//   - Prefix-based dictionary queries

#include <bits/stdc++.h>
using namespace std;

// Trie Node structure
struct TrieNode {
    bool endOfWord;
    unordered_map<char, TrieNode*> children;

    TrieNode() {
        endOfWord = false;
    }
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() {
        root = new TrieNode();
    }

    // Insert a word into the trie
    void insert(const string& word) {
        TrieNode* cur = root;
        for (char c : word) {
            if (!cur->children[c]) {
                cur->children[c] = new TrieNode();
            }
            cur = cur->children[c];
        }
        cur->endOfWord = true;
    }

    // Search if a complete word exists in the trie
    bool search(const string& word) {
        TrieNode* cur = root;
        for (char c : word) {
            if (!cur->children.count(c)) return false;
            cur = cur->children[c];
        }
        return cur->endOfWord;
    }

    // Check if any word starts with the given prefix
    bool startsWith(const string& prefix) {
        TrieNode* cur = root;
        for (char c : prefix) {
            if (!cur->children.count(c)) return false;
            cur = cur->children[c];
        }
        return true;
    }

    // Optional: delete word (advanced)
    bool deleteWord(const string& word) {
        return deleteHelper(root, word, 0);
    }

private:
    // Helper for deletion (removes nodes only if needed)
    bool deleteHelper(TrieNode* node, const string& word, int idx) {
        if (idx == word.size()) {
            if (!node->endOfWord) return false;
            node->endOfWord = false;
            return node->children.empty();
        }

        char c = word[idx];
        if (!node->children.count(c)) return false;

        bool shouldDelete = deleteHelper(node->children[c], word, idx + 1);

        if (shouldDelete) {
            delete node->children[c];
            node->children.erase(c);
            return !node->endOfWord && node->children.empty();
        }

        return false;
    }
};

// Example usage
int main() {
    Trie t;
    t.insert("apple");
    t.insert("app");

    cout << t.search("apple") << endl;     // 1
    cout << t.search("app") << endl;       // 1
    cout << t.search("ap") << endl;        // 0
    cout << t.startsWith("ap") << endl;    // 1

    return 0;
}
